sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeOptimizer.scala

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.{BoundReference, Expression}
+
+/**
+ * Optimizer to make decisions like, which expressions should we run on the CPU vs the GPU
+ * and combining cpu expressions into an expression tree.
+ */
+object GpuCpuBridgeOptimizer extends Logging {
+
+  def checkAndOptimizeExpressionMetas(exprs: Seq[BaseExprMeta[_]]): Unit = {
+    if (exprs.nonEmpty && exprs.head.conf.isCpuBridgeEnabled) {
+      exprs.foreach { child =>
+        if (!child.canExprTreeBeReplaced && canRunOnCpuOrGpuRecursively(child)) {
+          // Then we want to move some things to the CPU as needed
+          moveToCpuIfNeededRecursively(child, wasParentReplaced = false, 0)
+        }
+      }
+    }
+  }
+
+  private def canRunOnCpuOrGpuRecursively(expr: BaseExprMeta[_]): Boolean = {
+    if (!expr.canThisBeReplaced && !expr.canMoveToCpuBridge()) {
+      false
+    } else {
+      expr.childExprs.forall(canRunOnCpuOrGpuRecursively)
+    }
+  }
+
+  /**
+   * Move an expression to the GPU/CPU bridge if it has to or if it would be
+   * less data movement if it did run on the CPU, this means that it has more inputs
+   * and outputs on the CPU than the GPU.
+   * @param expr the expression to possibly move
+   * @param wasParentReplaced will its output be on the CPU
+   * @return true if this was placed on the CPU bridge else false
+   */
+  private def moveToCpuIfNeededRecursively(expr: BaseExprMeta[_],
+                                           wasParentReplaced: Boolean,
```

Comment:
I find it confusing that `wasParentReplaced` in this instance means it went to the CPU, whereas I believe `expr.canThisBeReplaced` is signifying whether it can run on the GPU.